staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go

```diff
@@ -67,17 +123,128 @@ var (
+	KeyIDHashLastTimestampSeconds = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "key_id_hash_last_timestamp_seconds",
+			Help:           "The last time in seconds when a keyID was used.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"transformation_type", "provider_name", "key_id_hash"},
+	)
+
+	// keyIDHashStatusLastTimestampSeconds is the last time in seconds when a keyID was returned by the Status RPC call.
+	// e.g. apiserver_envelope_encryption_key_id_hash_status_last_timestamp_seconds{key_id_hash="sha256", provider_name="providerName"} 1.674865558833728e+09
+	KeyIDHashStatusLastTimestampSeconds = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "key_id_hash_status_last_timestamp_seconds",
+			Help:           "The last time in seconds when a keyID was returned by the Status RPC call.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"provider_name", "key_id_hash"},
+	)
 )
 
 var registerMetricsFunc sync.Once
+var hashPool *sync.Pool
 
 func RegisterMetrics() {
+	randomCacheKey := make([]byte, 32)
+	keyIDHashTotalMetricLabels = &metricLabelsMap{
+		data: make(map[string]metricLabels),
+	}
+	keyIDHashStatusLastTimestampSecondsMetricLabels = &metricLabelsMap{
+		data: make(map[string]metricLabels),
+	}
+	hashPool = &sync.Pool{
+		New: func() interface{} {
+			return hmac.New(sha256.New, randomCacheKey)
+		},
+	}
 	registerMetricsFunc.Do(func() {
-		legacyregistry.MustRegister(dekCacheFillPercent)
-		legacyregistry.MustRegister(dekCacheInterArrivals)
+		legacyregistry.MustRegister(DekCacheFillPercent)
+		legacyregistry.MustRegister(DekCacheInterArrivals)
+		legacyregistry.MustRegister(KeyIDHashTotal)
+		legacyregistry.MustRegister(KeyIDHashLastTimestampSeconds)
+		legacyregistry.MustRegister(KeyIDHashStatusLastTimestampSeconds)
 	})
 }
 
+// RecordKeyID records total count and last time in seconds when a KeyID was used for TransformFromStorage and TransformToStorage operations
+func RecordKeyID(transformationType, providerName, keyID string) {
+	h := hashPool.Get().(hash.Hash)
+	h.Reset()
+	h.Write([]byte(keyID))
+	hashPool.Put(h)
+	keyIDHash := fmt.Sprintf("sha256:%x", h.Sum(nil))
+
+	mapKey := providerName + keyIDHash
+	if !keyIDHashTotalMetricLabels.exists(mapKey) {
+		keyIDHashTotalMetricLabels.addItem(mapKey, metricLabels{
+			transformationType: transformationType,
+			providerName:       providerName,
+			keyIDHash:          keyIDHash,
+		})
+		if len(keyIDHashTotalMetricLabels.data) > maxKeyIDLabelsLimit {
+			item, err := keyIDHashTotalMetricLabels.getFirstItem()
+			if err != nil {
+				return
+			}
+			KeyIDHashTotal.DeleteLabelValues(item.transformationType, item.providerName, item.keyIDHash)
+			KeyIDHashLastTimestampSeconds.DeleteLabelValues(item.transformationType, item.providerName, item.keyIDHash)
+			keyIDHashTotalMetricLabels.deleteFirstItem()
```

Comment:
A unit test for this functionality would be great.. i.e. record N+1 different values and verify the output for each.

Reply:
I've verified the expected behavior as part of `TestEnvelopeMetrics` in https://github.com/kubernetes/kubernetes/pull/115394/files#diff-7453383257a423497b7e2893ec1fec0b7baa6fd55a3d7e228aa5da475b9ced74R40 by adding different keyIDs and restricting `maxKeyIDLabelsLimit` to 2. lmk if that is sufficient.

Reply:
Ah, I see. Yeah that looks great! Thanks!